src/lib.rs

```diff
@@ -470,200 +469,891 @@ pub struct AtLeast {
+            } else {
+                res0 = lineqs0.iter().any(|x| x.satisfied(variable_combination.clone()));
+                res1 = lineqs1.iter().any(|x| x.satisfied(variable_combination.clone()));
+            }
+            return (res0 & res1) | !(res0 & res1);
+        })
+    }
+
+    fn validate_theory_lineqs(t: Theory, actual: Vec<GeLineq>, expected: Vec<GeLineq>) -> bool {
+        return validate_all_combinations(
+            t.statements.into_iter().map(|x| (x.variable.id, x.variable.bounds)).collect(), 
+            actual, 
+            expected, 
+            true
+        )
+    }
+
+    #[test]
+    fn test_theory_to_lineqs_reduced() {
+
+        // Depth 3
+        // 0: 1 & 2
+        // 1: 3 + 4 >= 6 (3 has bounds (-5,5), 4 is bool) 
+        // 2: 5 + 6 >= 4 (5 has bounds (-3,3), 6 is bool) 
+
+        // Expected constraints:
+        // 4( x) +( y) -41
+        let t = Theory {
+            id: String::from("A"),
+            statements: vec![
+                Statement {
+                    variable: Variable { id: 0, bounds: (0,1) },
+                    expression: Some(
+                        AtLeast {
+                            ids: vec![1,2],
+                            bias: -2
+                        }
+                    )
+                },
+                Statement {
+                    variable: Variable { id: 1, bounds: (0,1) },
+                    expression: Some(
+                        AtLeast {
+                            ids: vec![3,4],
+                            bias: -6
+                        }
+                    )
+                },
+                Statement {
+                    variable: Variable { id: 2, bounds: (0,1) },
+                    expression: Some(
+                        AtLeast {
+                            ids: vec![5,6],
+                            bias: -4
+                        }
+                    )
+                },
+                Statement {
+                    variable: Variable { id: 3, bounds: (-5,5) },
+                    expression: None
+                },
+                Statement {
+                    variable: Variable { id: 4, bounds: (0,1) },
+                    expression: None
+                },
+                Statement {
+                    variable: Variable { id: 5, bounds: (-3,3) },
+                    expression: None
+                },
+                Statement {
+                    variable: Variable { id: 6, bounds: (0,1) },
+                    expression: None
+                },
+            ]
+        };
+        let actual: Vec<GeLineq> = t.to_lineqs(true, true);
+        let expected: Vec<GeLineq> = vec![
+            GeLineq {
+                bias: -34,
+                bounds: vec![(-5,5),(0,1),(-3,3),(0,1)],
+                coeffs: vec![5,5,1,1],
+                indices: vec![3,4,5,6]
+            },
+        ];
+        assert!(validate_theory_lineqs(t, actual, expected));
```

Comment:
I believe we want to validate the theory based on the input and actual and not on the expected